include/nlohmann/detail/output/serializer.hpp

```diff
@@ -104,272 +104,320 @@ class serializer
     @param[in] indent_step       the indent level
     @param[in] current_indent    the current indent level (only used internally)
     */
-    void dump(const BasicJsonType& val,
+    void dump(const BasicJsonType& start_val,
               const bool pretty_print,
               const bool ensure_ascii,
               const unsigned int indent_step,
-              const unsigned int current_indent = 0)
+              const unsigned int start_indent = 0)
     {
-        switch (val.m_data.m_type)
+        struct dump_state
         {
-            case value_t::object:
-            {
-                if (val.m_data.m_value.object->empty())
-                {
-                    o->write_characters("{}", 2);
-                    return;
-                }
+            const BasicJsonType& val;
+            const unsigned int new_indent;
+            const unsigned int offset;
+            const decltype(start_val.m_data.m_value.object->cbegin()) iterator;
+        };
+        std::vector<dump_state> stack;
+        stack.push_back({start_val, start_indent, 0});
 
-                if (pretty_print)
+        while (!stack.empty())
+        {
+            const dump_state& state = stack.back();
```

Comment:
By taking a reference here, you end up with a dangling reference when you do `pop_back()`.  Even if you don't use `state` again after the `pop_back()`, you are referring to the `BasicJsonType` inside it with `val`.  It's probably better to store a `BasicJsonType *` in the state, so you don't have to worry about referenceds to references, and then you can do this:

    const BasicJsonType* val = stack.back().val;
    const unsigned int current_indent = stack.back().new_indent;
    const unsigned int offset = stack.back().offset;
    auto iterator = stack.back().iterator;
    stack.pop_back();